AX: [iOS] Date fields shouldn't send focused change notification when picker popup is presented

hoffmanjoshua opened the pull request:
#### 8c8a97961d8bdac770f70e168b00693feffccdc3
<pre>
AX: [iOS] Date fields shouldn&apos;t send focused change notification when picker popup is presented
<a href="https://bugs.webkit.org/show_bug.cgi?id=269269">https://bugs.webkit.org/show_bug.cgi?id=269269</a>
<a href="https://rdar.apple.com/122851275">rdar://122851275</a>

Reviewed by Tyler Wilcock.

VoiceOver focus wasn&apos;t being moved to the date picker&apos;s popup due to a focused element changed
notification fired on the date element as the picker was expanded. Native pop up elements will
automatically post this notification in the UI process, so this notification conflicted and
created a race condition.

To prevent this, there is a new flag in AXObjectCache that maintains the state of the picker.
If the picker will open, this flag prevents us from sending the conflicting notification. We
clear this state for new focus changes and when the picker is closed (sent via IPC from the
UI process).

Date and DateTimeLocal input types were also added to elements that should defer their focus
(shouldDeferFocusChange). This is because when this race condition occurs, the focus event
happens prior to the DOMActivate event being dispatched, which prevents us from knowing the
state of the picker before handling the notification. By deferring, we allow the DOM event
to occur in time.

Test: accessibility/ios-simulator/date-picker-focus-notification.html

* LayoutTests/accessibility/ios-simulator/date-picker-focus-notification-expected.txt: Added.
* LayoutTests/accessibility/ios-simulator/date-picker-focus-notification.html: Added.
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::shouldDeferFocusChange):
(WebCore::AXObjectCache::handleFocusedUIElementChanged):
* Source/WebCore/accessibility/AXObjectCache.h:
* Source/WebCore/html/BaseDateAndTimeInputType.cpp:
(WebCore::BaseDateAndTimeInputType::showPicker):
* Source/WebCore/page/Page.cpp:
(WebCore::Page::clearIsShowingInputView):
* Source/WebCore/page/Page.h:
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::setIsShowingInputViewForFocusedElement):

Canonical link: <a href="https://commits.webkit.org/301057@main">https://commits.webkit.org/301057@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/07ba4e51b66034785b31bf2de74e805274b0de9c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/124722 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/44394 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/35129 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/131571 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/76643 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/c9b3b021-e7c0-4ab0-9b1e-4b71e95ff377) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/126599 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/45098 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/52964 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/94896 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/62950 "Passed tests") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/9b2aaa62-f639-4605-b885-7ea53ff6416a) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/127676 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/35971 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/111546 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/75465 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/087a813b-6ff3-49be-a67c-09d4cdbb7b91) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/34902 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/29700 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/75049 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/105728 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/29931 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/134237 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/51570 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/39385 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/103372 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/51983 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/107766 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/103145 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26284 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/48509 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/26788 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/48558 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/51443 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/57241 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/50836 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/54192 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/52531 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

minorninth commented on Source/WebCore/accessibility/AXObjectCache.cpp:
```diff
@@ -1753,6 +1753,16 @@ static bool shouldDeferFocusChange(Element* element)
+#if PLATFORM(IOS_FAMILY)
+    // Date/DateTimeLocal input fields present popovers in the UI process synchronously.
+    // The web process can infer that this will happen via a DOMActivateEvent, but this may be dispatched after the focused change event.
+    // Defer the focus so that we can process the DOM activate event and set the appropriate state (willPresentDatePopover) on AXObjectCache.
```
I was confused because I thought "we" here meant AXObjectCache processing the DOM activate event. Maybe you could explicitly say: defer focus until after BaseDateAndTimeInputType::showPicker is called.

hoffmanjoshua replied:
Great call! Updated in the latest commit

minorninth reviewed (commented):
lgtm with one clarification request!

twilco commented on Source/WebCore/page/Page.cpp:
```diff
@@ -5971,6 +5971,14 @@ RefPtr<MediaSessionManagerInterface> Page::mediaSessionManagerForPageIdentifier(
+void Page::clearIsShowingInputView()
+{
+    if (CheckedPtr axObjectCache = existingAXObjectCache())
+        axObjectCache->setWillPresentDatePopover(false);
```
Maybe just `cache`? We use that in most places

hoffmanjoshua replied:
Fixed in latest commit

twilco commented on Source/WebCore/html/BaseDateAndTimeInputType.cpp:
```diff
@@ -341,13 +341,20 @@ void BaseDateAndTimeInputType::showPicker()
 
-    if (!element()->document().page())
+    RefPtr document = element()->document();
+    if (!document->page())
```
Do we need a null-check on document?

hoffmanjoshua replied:
This can actually be a Ref because document is a reference. Updated

twilco commented on LayoutTests/accessibility/ios-simulator/date-picker-focus-notification.html:
```diff
@@ -0,0 +1,43 @@
+
+</script>
+
+
```
Extra newlines

twilco commented on LayoutTests/accessibility/ios-simulator/date-picker-focus-notification.html:
```diff
@@ -0,0 +1,43 @@
+		finishJSTest();
+	}, 0);
+}
+
```
Extra newline

twilco commented on LayoutTests/accessibility/ios-simulator/date-picker-focus-notification.html:
```diff
@@ -0,0 +1,43 @@
+		datefield.takeFocus();
+
+		//  We shouldn't receive any focused notifications for this date picker. 
+		await sleep(0.5);
```
Is sleeping for half a millisecond what you intended here?

hoffmanjoshua replied:
The bug still reproduces (without my change) with this short of a wait, but we can bump this to 250ms to prevent any flakeyness that would prevent us from catching bugs in the future.

twilco replied:
250ms is probably too long, maybe we go for 60ms?

twilco commented on LayoutTests/accessibility/ios-simulator/date-picker-focus-notification.html:
```diff
@@ -0,0 +1,43 @@
+	datefield.addNotificationListener(focusCallback);
+	
+	setTimeout(async function() {
+		await UIHelper.activateElement(document.getElementById('datefield'));
```
For better or for worse, we have landed on using double quotes for strings in our tests

twilco commented on LayoutTests/accessibility/ios-simulator/date-picker-focus-notification.html:
```diff
@@ -0,0 +1,43 @@
+function focusCallback(notification) {
+    if (notification === "AXFocusChanged") {
+    	output += "Received AXFocusChanged event\n";
+    }
```
Curly braces aren't necessary here

twilco commented on LayoutTests/accessibility/ios-simulator/date-picker-focus-notification.html:
```diff
@@ -0,0 +1,43 @@
+
+
+<script>
+var output = "this tests that when a date field shows a popover, no focused event is fired.\n\n";
```
Missing capital letter at the start of this sentence

twilco commented on LayoutTests/accessibility/ios-simulator/date-picker-focus-notification.html:
```diff
@@ -0,0 +1,43 @@
+
+<input id="datefield" type="date">
+
+
```
Extra newline

twilco commented on LayoutTests/accessibility/ios-simulator/date-picker-focus-notification.html:
```diff
@@ -0,0 +1,43 @@
+<head>
+<script src="../../resources/js-test.js"></script>
+<script src="../../resources/accessibility-helper.js"></script>
+<script src="../../resources/ui-helper.js"></script>
```
`accessiblity-helper.js` should be included first as it comes first alphabetically

hoffmanjoshua replied:
Fixed

twilco commented on Source/WebCore/accessibility/AXObjectCache.h:
```diff
@@ -871,6 +874,10 @@ class AXObjectCache final : public CanMakeWeakPtr<AXObjectCache>, public CanMake
+
+#if PLATFORM(IOS_FAMILY)
+    bool m_willPresentDatePopover;
+#endif
```
Is this placed into padding?

hoffmanjoshua replied:
Moved to fit into some padding above.

twilco commented on LayoutTests/accessibility/ios-simulator/date-picker-focus-notification.html:
```diff
@@ -0,0 +1,39 @@
+		await UIHelper.activateElement(document.getElementById("datefield"));
+		datefield.takeFocus();
+
+		//  We shouldn't receive any focused notifications for this date picker. 
```
Extra space after the //

twilco replied:
Maybe the comment can explicitly state this is the reason for the wait?

`// We shouldn't receive any focused notifications for this date picker, so wait some time to allow the bug to occur if our implementation is wrong.`

hoffmanjoshua committed:
AX: [iOS] Date fields shouldn't send focused change notification when picker popup is presented
https://bugs.webkit.org/show_bug.cgi?id=269269
rdar://122851275

Reviewed by Tyler Wilcock.

VoiceOver focus wasn't being moved to the date picker's popup due to a focused element changed
notification fired on the date element as the picker was expanded. Native pop up elements will
automatically post this notification in the UI process, so this notification conflicted and
created a race condition.

To prevent this, there is a new flag in AXObjectCache that maintains the state of the picker.
If the picker will open, this flag prevents us from sending the conflicting notification. We
clear this state for new focus changes and when the picker is closed (sent via IPC from the
UI process).

Date and DateTimeLocal input types were also added to elements that should defer their focus
(shouldDeferFocusChange). This is because when this race condition occurs, the focus event
happens prior to the DOMActivate event being dispatched, which prevents us from knowing the
state of the picker before handling the notification. By deferring, we allow the DOM event
to occur in time.

Test: accessibility/ios-simulator/date-picker-focus-notification.html

* LayoutTests/accessibility/ios-simulator/date-picker-focus-notification-expected.txt: Added.
* LayoutTests/accessibility/ios-simulator/date-picker-focus-notification.html: Added.
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::shouldDeferFocusChange):
(WebCore::AXObjectCache::handleFocusedUIElementChanged):
* Source/WebCore/accessibility/AXObjectCache.h:
* Source/WebCore/html/BaseDateAndTimeInputType.cpp:
(WebCore::BaseDateAndTimeInputType::showPicker):
* Source/WebCore/page/Page.cpp:
(WebCore::Page::clearIsShowingInputView):
* Source/WebCore/page/Page.h:
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::setIsShowingInputViewForFocusedElement):

Canonical link: https://commits.webkit.org/301057@main
